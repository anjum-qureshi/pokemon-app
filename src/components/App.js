import React, { useEffect, useState } from 'react';

const PokemonThumb = ({ image, name, type, showDetails }) => {
  return (
    <div className={`${type} thumb-container`} onClick={showDetails} data-testid="pokemon-tile">
      <img src={image} alt={name} />
      <div className="detail-wrapper">
        <h3>{name}</h3>
      </div>
    </div>
  );
};

const url = 'https://pokeapi.co/api/v2/pokemon';

const App = () => {
  const [allPokemons, setAllPokemons] = useState([]);
  const [visible, setVisible] = useState(false);
  const [loadMore, setLoadMore] = useState(`${url}?limit=2`);
  const [currIndex, setCurrIndex] = useState(null);

  const getAllPokemons = async () => {
    const resData = await fetch(loadMore);
    const response = await resData.json();
    setLoadMore(response.next);

    const pokemons = await Promise.all(
      response.results.map(async (pokemon) => {
        return await (await fetch(`${url}/${pokemon.name}`)).json();
      })
    );
    const sortedPokemons = [...allPokemons, ...pokemons];
    sortedPokemons.sort((a, b) => a.id - b.id);
    setAllPokemons(sortedPokemons);
  };

  useEffect(() => {
    getAllPokemons();
  }, []);

  const currentPokemon = allPokemons[currIndex];

  return (
    <div className="app-contaner">
      <h1>Pokemon</h1>
      <div className="pokemon-container">
        <div className="all-container">
          {allPokemons.map((p, index) => {
            return (
              <div
                key={`pokemon-${index}`}
                className={`${p.types[0].type.name} thumb-container`}
                onClick={() => {
                  setCurrIndex(index);
                  setVisible(true);
                }}
                data-testid="pokemon-tile"
              >
                <img src={p.sprites.other.dream_world.front_default} alt={p.name} />
                <div className="detail-wrapper">
                  <h3>{p.name}</h3>
                </div>
              </div>
              // <PokemonThumb
              //   key={index}
              //   image={p.sprites.other.dream_world.front_default}
              //   name={p.name}
              //   type={p.types[0].type.name}
              // showDetails={() => {
              //   setCurrIndex(index);
              //   setVisible(true);
              // }}
              //   hideDetails={() => setVisible(false)}
              // />
            );
          })}
        </div>
        <button className="load-more" onClick={() => getAllPokemons()}>
          Load More
        </button>
      </div>
      {visible && (
        <div className={`popup ${visible ? 'show' : 'hide'}`}>
          <h2>{currentPokemon.name}</h2>
          <img
            src={`https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/${currentPokemon.id}.svg`}
            alt=""
          />
          <div className="abilities">
            {currentPokemon.abilities.map((poke, idx) => (
              <div className="group" key={`group-poke-${idx}`}>
                <h2>{poke.ability.name}</h2>
              </div>
            ))}
          </div>
          <div className="base-stat">
            {currentPokemon.stats.map((poke) => {
              return (
                <h3>
                  {poke.stat.name}:{poke.base_stat}
                </h3>
              );
            })}
          </div>
          <button className="load-more" onClick={() => setVisible(false)}>
            close
          </button>
        </div>
      )}
    </div>
  );
};

export default App;
