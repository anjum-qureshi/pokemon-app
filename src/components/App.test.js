import React from 'react';
import { render, act, waitFor } from '@testing-library/react';
import App from './App';
import { allPokemonsResponse, pokemonData } from './testData';

describe('App test', () => {
  beforeEach(() => {
    global.fetch = jest
      .fn()
      .mockImplementationOnce(() => {
        return Promise.resolve({
          json: () => {
            return Promise.resolve(allPokemonsResponse);
          },
        });
      })
      .mockImplementation(() => {
        return Promise.resolve({
          json: () => {
            return Promise.resolve(pokemonData);
          },
        });
      });
  });

  afterEach(() => {
    global.fetch.mockClear();
    jest.restoreAllMocks();
  });

  it('Generate snapshot for App', async () => {
    const container = await render(<App />);
    await act(async () => {});
    expect(container).toMatchSnapshot();
    // const { getAllByTestId } = container;

    // await waitFor(async () => {
    //   const tiles = await getAllByTestId('pokemon-tile');
    //   expect(tiles).toHaveLength(2);
    // });
  });

  //   it('Should render first two pokemons when app is rendered', async () => {
  //     const container = await render(<App />);
  //     await act(async () => {});

  //     const { getAllByTestId, queryAllByTestId } = container;
  //     const tiles = await getAllByTestId('pokemon-tile');
  //     expect(tiles).toHaveLength(2);
  //   });
});
