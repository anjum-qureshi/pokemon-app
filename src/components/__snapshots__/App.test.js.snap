// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App test Generate snapshot for App 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="app-contaner"
      >
        <h1>
          Pokemon
        </h1>
        <div
          class="pokemon-container"
        >
          <div
            class="all-container"
          >
            <div
              class="grass thumb-container"
              data-testid="pokemon-tile"
            >
              <img
                alt="bulbasaur"
                src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg"
              />
              <div
                class="detail-wrapper"
              >
                <h3>
                  bulbasaur
                </h3>
              </div>
            </div>
          </div>
          <button
            class="load-more"
          >
            Load More
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="app-contaner"
    >
      <h1>
        Pokemon
      </h1>
      <div
        class="pokemon-container"
      >
        <div
          class="all-container"
        >
          <div
            class="grass thumb-container"
            data-testid="pokemon-tile"
          >
            <img
              alt="bulbasaur"
              src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg"
            />
            <div
              class="detail-wrapper"
            >
              <h3>
                bulbasaur
              </h3>
            </div>
          </div>
        </div>
        <button
          class="load-more"
        >
          Load More
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
